File: src/DungeonDelverC/Boilerplate.c
#include "Globals.h"
#include "Boilerplate.h"
#include "Application.h"
struct Boilerplate
{
	int running;
	struct Globals* globals;
	struct Application* application;
};
struct Globals* Boilerplate_GetGlobals(struct Boilerplate* ptr)
{
	return ptr->globals;
}
struct Application* Boilerplate_GetApplication(struct Boilerplate* ptr)
{
	return ptr->application;
}
static char Boilerplate_GetCommand(SDL_KeyCode code)
{
	switch (code)
	{
	case SDLK_INSERT:
	case SDLK_KP_0:
	case SDLK_0:
		return '0';
	case SDLK_END:
	case SDLK_KP_1:
	case SDLK_1:
		return '1';
	case SDLK_DOWN:
	case SDLK_KP_2:
	case SDLK_2:
		return '2';
	case SDLK_KP_3:
	case SDLK_3:
	case SDLK_PAGEDOWN:
		return '3';
	case SDLK_KP_4:
	case SDLK_LEFT:
	case SDLK_4:
		return '4';
	case SDLK_KP_5:
	case SDLK_5:
		return '5';
	case SDLK_KP_6:
	case SDLK_6:
	case SDLK_RIGHT:
		return '6';
	case SDLK_HOME:
	case SDLK_KP_7:
	case SDLK_7:
		return '7';
	case SDLK_UP:
	case SDLK_KP_8:
	case SDLK_8:
		return '8';
	case SDLK_KP_9:
	case SDLK_PAGEUP:
	case SDLK_9:
		return '9';
	case SDLK_KP_PERIOD:
	case SDLK_PERIOD:
	case SDLK_DELETE:
		return '.';
	default:
		return 0;
	}
}
void Boilerplate_Loop(struct Boilerplate* ptr)
{
	SDL_Event event = { 0 };
	for (;ptr->running;)
	{
		SDL_RenderClear(Globals_GetRenderer(Boilerplate_GetGlobals(ptr)));
		Application_Draw(Boilerplate_GetApplication(ptr));
		SDL_RenderPresent(Globals_GetRenderer(Boilerplate_GetGlobals(ptr)));
		while(ptr->running && SDL_PollEvent(&event))
		{
			switch (event.type)
			{
			case SDL_QUIT:
				ptr->running = 0;
				break;
			case SDL_KEYDOWN:
				ptr->running = Application_HandleCommand(ptr->application, Boilerplate_GetCommand(event.key.keysym.sym));
				break;
			}
		}
	}
}
struct Boilerplate* Boilerplate_Initialize()
{
	if (0 != SDL_Init(SDL_INIT_EVERYTHING))
	{
		return 0;
	}
	struct Boilerplate* ptr = SDL_malloc(sizeof(struct Boilerplate));
	SDL_memset(ptr, 0, sizeof(struct Boilerplate));
	ptr->globals = Globals_Initialize();

	SDL_Window* window = 0;
	SDL_Renderer* renderer = 0;
	int running = 0 == SDL_CreateWindowAndRenderer(SCREEN_WIDTH, SCREEN_HEIGHT, 0, &window, &renderer);
	Globals_SetRenderer(ptr->globals, renderer);
	Globals_SetWindow(ptr->globals, window);
	ptr->application = Application_Initialize(ptr->globals);
	ptr->running = 1;
	return ptr;
}
void Boilerplate_CleanUp(struct Boilerplate** ptr)
{
	Application_CleanUp(&(*ptr)->application);
	if (Globals_GetRenderer(Boilerplate_GetGlobals(*ptr)))
	{
		SDL_DestroyRenderer(Globals_GetRenderer(Boilerplate_GetGlobals(*ptr)));
		Globals_SetRenderer(Boilerplate_GetGlobals(*ptr), 0);
	}
	if (Globals_GetWindow(Boilerplate_GetGlobals(*ptr)))
	{
		SDL_DestroyWindow(Globals_GetWindow(Boilerplate_GetGlobals(*ptr)));
		Globals_SetWindow(Boilerplate_GetGlobals(*ptr), 0);
	}
	Globals_CleanUp(&(*ptr)->globals);
	SDL_Quit();
}

File: src/DungeonDelverVB/Boilerplate.vb
Module Boilerplate
    Sub CleanUp(cleaner As Action)
        cleaner()
        If g_renderer <> IntPtr.Zero Then
            SDL_DestroyRenderer(g_renderer)
            g_renderer = IntPtr.Zero
        End If
        If g_window <> IntPtr.Zero Then
            SDL_DestroyWindow(g_window)
            g_window = IntPtr.Zero
        End If
        SDL_Quit()
    End Sub
    Function Initialize(initializer As Action) As Boolean
        If 0 <> SDL_Init(&H10F231) Then
            Return False
        End If
        Dim running As Boolean = 0 = SDL_CreateWindowAndRenderer(SCREEN_WIDTH, SCREEN_HEIGHT, 0, g_window, g_renderer)
        If running Then
            initializer()
        End If
        Return running
    End Function
    Sub [Loop](running As Boolean, drawer As Action, updater As Func(Of Char, Boolean))
        Dim [event] As New SDL_Event
        Do
            SDL_RenderClear(g_renderer)
            SDL_RenderPresent(g_renderer)
            If SDL_WaitEvent([event]) = 1 Then
                Select Case [event].type
                    Case 256
                        Exit Do
                    Case 768
                        If Not HandleKeyDown([event].sym, updater) Then
                            Exit Do
                        End If
                End Select
            End If
        Loop
    End Sub
    ReadOnly keyTable As IReadOnlyDictionary(Of UInt32, Char) =
        New Dictionary(Of UInt32, Char) From
        {
             {SDLK_PERIOD, COMMAND_DOT},
             {SDLK_0, COMMAND_ZERO},
             {SDLK_1, COMMAND_ONE},
             {SDLK_2, COMMAND_TWO},
             {SDLK_3, COMMAND_THREE},
             {SDLK_4, COMMAND_FOUR},
             {SDLK_5, COMMAND_FIVE},
             {SDLK_6, COMMAND_SIX},
             {SDLK_7, COMMAND_SEVEN},
             {SDLK_8, COMMAND_EIGHT},
             {SDLK_9, COMMAND_NINE},
             {SDLK_DELETE, COMMAND_DOT},
             {SDLK_INSERT, COMMAND_ZERO},
             {SDLK_HOME, COMMAND_SEVEN},
             {SDLK_PAGEUP, COMMAND_NINE},
             {SDLK_END, COMMAND_ONE},
             {SDLK_PAGEDOWN, COMMAND_THREE},
             {SDLK_RIGHT, COMMAND_SIX},
             {SDLK_LEFT, COMMAND_FOUR},
             {SDLK_DOWN, COMMAND_TWO},
             {SDLK_UP, COMMAND_EIGHT},
             {SDLK_KP_1, COMMAND_ONE},
             {SDLK_KP_2, COMMAND_TWO},
             {SDLK_KP_3, COMMAND_THREE},
             {SDLK_KP_4, COMMAND_FOUR},
             {SDLK_KP_5, COMMAND_FIVE},
             {SDLK_KP_6, COMMAND_SIX},
             {SDLK_KP_7, COMMAND_SEVEN},
             {SDLK_KP_8, COMMAND_EIGHT},
             {SDLK_KP_9, COMMAND_NINE},
             {SDLK_KP_0, COMMAND_ZERO},
             {SDLK_KP_PERIOD, COMMAND_DOT}
        }
    Private Function HandleKeyDown(sym As UInteger, updater As Func(Of Char, Boolean)) As Boolean
        Dim character As Char = Char.MinValue
        If keyTable.TryGetValue(sym, character) Then
            Return updater(character)
        End If
        Return True
    End Function
End Module
